ouroboros-consensus/test-consensus/Test/Dynamic/Ref/RealPBFT.hs

```diff
@@ -0,0 +1,255 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | A reference simulator of the RealPBFT protocol under \"ideal
+-- circumstances\"
+--
+-- See 'step'.
+--
+module Test.Dynamic.Ref.RealPBFT (
+  Outcome (..),
+  State (..),
+  advanceUpTo,
+  emptyState,
+  nullState,
+  step,
+  viable,
+  ) where
+
+import           Data.Foldable (Foldable, foldl', toList)
+import qualified Data.Map as Map
+import           Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+
+import           Ouroboros.Network.Block (SlotNo (..))
+
+import           Ouroboros.Consensus.NodeId (CoreNodeId (..))
+import           Ouroboros.Consensus.Protocol.Abstract (SecurityParam (..))
+import           Ouroboros.Consensus.Protocol.PBFT (PBftParams (..))
+
+import           Test.Dynamic.Util.NodeJoinPlan
+
+oneK :: Num a => PBftParams -> a
+oneK PBftParams{pbftSecurityParam} =
+    fromIntegral (maxRollbacks pbftSecurityParam)
+
+twoK :: Num a => PBftParams -> a
+twoK PBftParams{pbftSecurityParam} =
+    2 * fromIntegral (maxRollbacks pbftSecurityParam)
+
+oneN :: Num a => PBftParams -> a
+oneN PBftParams{pbftNumNodes} = fromIntegral pbftNumNodes
+
+{-------------------------------------------------------------------------------
+  PBFT state
+-------------------------------------------------------------------------------}
+
+-- | Outcome of a node leading a slot
+--
+data Outcome
+
+  = Absent
+    -- ^ the leader hadn't yet joined
+
+  | Nominal
+    -- ^ the leader extended the networks' common chain with a valid block
+
+  | Unable
+    -- ^ the leader could not forge a valid block
+
+  | Wasted
+    -- ^ the leader forged a valid-but-irrelevant block, because it forged
+    -- before sufficiently synchronizing
+    --
+    -- As of commit 4eb23cfe \"Merge #1260\", this only happens if a node leads
+    -- in the slot it joins. In that slot, it forges before the mini protocols
+    -- are able to pull in any of the existing nodes' chain, and so it
+    -- re-forges the epoch 0 boundary block (EBB).
+
+  deriving (Eq, Show)
+
+-- | The state of the RealPBFT net
+--
+data State = State
+  { ids      :: !(Seq CoreNodeId)
```

Comment:
Maybe `forgers` would be a more self-explanatory name?

Reply:
Fixed, thanks.